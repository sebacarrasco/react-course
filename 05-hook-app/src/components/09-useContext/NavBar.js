import React from 'react';
import { NavLink } from 'react-router-dom';

// La diferencia entre Link y NavLink es que navlink acepta el atributo
// activeClassName, que permita agregar una clase al link cuando
// su ruta estÃ¡ activa.

export const NavBar = () => {
    return (

        <nav className="navbar navbar-expand-lg navbar-dark bg-dark">
            <div className="container-fluid">

                <NavLink className="navbar-brand" to="/">Usecontext App</NavLink>

                <div className="collapse navbar-collapse" id="navbarNav">
                    
                    <div className="navbar-nav">
                        <NavLink className="nav-item nav-link" activeClassName="active" exact to="/">Home</NavLink>
                    
                        <NavLink className="nav-item nav-link" activeClassName="active" exact to="./about">About</NavLink>
                    
                        <NavLink className="nav-item nav-link" activeClassName="active" exact to="./login">Login</NavLink>
                    </div>
                </div>

            </div>
        </nav>
    )
}
