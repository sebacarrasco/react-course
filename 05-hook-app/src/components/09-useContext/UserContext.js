import { createContext } from "react";

// Sirve para compartir información entre varios componentes que
// no necesariamente son hijos, sino que pueden ser hermanos también
export const UserContext = createContext(null);