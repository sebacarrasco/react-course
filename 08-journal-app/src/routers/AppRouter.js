import { firebase } from '../firebase/firebase-config';
import React, { useEffect, useState } from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Redirect
  } from "react-router-dom";
import { JournalScreen } from '../components/journal/JournalScreen';
import { AuthRouter } from './AuthRouter';
import { useDispatch } from 'react-redux';
import { login } from '../actions/auth';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';
import { startLoadingNotes } from '../actions/notes';

export const AppRouter = () => {

    const dispatch = useDispatch();

    // Chequeando el estado de firebase
    const [checking, setChecking] = useState(true);

    const [isLogged, setIsLogged] = useState(false);

    // Este useEffect hace que no se elimine el estado de loggeado al
    // recargar la página. Esto es debido a que cuando se refresque la página
    // el onAuthStateChanged se ejecuta (haciendo que nos subscribamos -> estamos añadiendonos como observadores),
    // entonces en caso de que el usuario haya estado loggeado, este se recupera y
    // podemos actualizar la store (Creo que firebase guarda al usuario en indexDb (ver en devtools) y por eso no se pierde
    // al recargar).
    // Estamos diciendo que esto se ejecute justo después de que se cargue la página.
    useEffect(() => {
        // Esto se realiza de maner asíncrona, por eso usamos el checking
        firebase.auth().onAuthStateChanged( async(user) => {
            // Esta función se ejecutará cada vez que cambie la autenticación del usuario
            // y cuando se recarga la página
            // Estamos subscribiendo esta función a los cambios dela auth (observable)

            // Si no está autenticado, user será null
            // Si existe el uid, el user está autenticado
            if (user?.uid)
            {  
                // Si hay user guardado en indexDb, entonces lo loggeamos
                dispatch(login(user.uid, user.displayName));
                setIsLogged(true);

                dispatch(startLoadingNotes(user.uid));
            }
            else
            {
                setIsLogged(false);
            }
            setChecking(false);
    
        });
    }, [ dispatch, setChecking, setIsLogged ])
    // Esto solo se ejecutará una vez porque ni dispatch ni el setChecking cambiarán


    // mientras se está chequeando si el user sigue loggeado, se muestra esto.
    if (checking)
    {
        return (
            <h1 style={ {textAlign:"center"} }>
                Loading...
            </h1>
        )
    }


    // Cuando el checking cambia a false y ya sabemos si tenemos un usuario loggeado o no,
    // entonces retornamos el router
    return (
        <Router>
            <div>
                <Switch>
                    <PublicRoute path="/auth" component={ AuthRouter } isLogged={ isLogged }/>

                    <PrivateRoute exact path="/" component={ JournalScreen } isLogged={ isLogged }/>

                    <Redirect to="/auth/login" />
                </Switch>
            </div>
        </Router>
    )
}
