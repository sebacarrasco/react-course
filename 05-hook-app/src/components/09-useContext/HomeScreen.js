import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const HomeScreen = () => {

    // Le pido que encuentre la instancia de UserContext en el Ã¡rbol
    const { user } = useContext(UserContext)
    
    return (
        <div>
            <h1> Home Screen </h1>
            <hr />

            <pre className="container">
                { JSON.stringify( user, null, 3) }
            </pre>
        </div>
    )
}
