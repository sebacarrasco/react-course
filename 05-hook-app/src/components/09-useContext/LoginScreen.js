import React, { useContext } from 'react'
import { UserContext } from './UserContext'

export const LoginScreen = () => {

    const { setUser } = useContext(UserContext);
    const user = {
        id: 123,
        name: "Seba"
    }

    return (
        <div>
            <h1> Login Screen </h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={ () => setUser(user) }
            >
                Login
            </button>
        </div>
    )
}
