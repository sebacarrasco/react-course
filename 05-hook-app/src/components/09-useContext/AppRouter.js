import React from 'react';
import {
    BrowserRouter as Router,
    Switch,
    Route,
    Redirect
  } from "react-router-dom";
import { AboutScreen } from './AboutScreen';
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';
import { NavBar } from './NavBar';

export const AppRouter = () => {
    return (
        <Router>
            <div>

                <NavBar />

                <div className="container">
                    <Switch>

                        {/* El exact es para que sea exactamente ese path */}
                        <Route exact path="/" component={ HomeScreen } />
                        <Route exact path="/about" component={ AboutScreen } />
                        <Route exact path="/login" component={ LoginScreen } />

                        {/* Nótese que no se le pone path, es decir, si se pone una ruta
                        que no calza con ninguna de las anteriores, entonces se mandará a la homescreen */}
                        {/* <Route component={ HomeScreen } /> */}

                        {/* Lo de arriba es equivalente a hacer: */}
                        <Redirect to="/" />

                    </Switch>
                </div>
            </div>
        </Router>
    )
}
