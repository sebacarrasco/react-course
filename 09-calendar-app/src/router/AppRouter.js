import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
    BrowserRouter as Router,
    Switch,
    Redirect
  } from "react-router-dom";
import { startChecking } from '../actions/auth';
import { LoginScreen } from '../components/auth/LoginScreen';
import { CalendarScreen } from '../components/calendar/CalendarScreen';
import { PrivateRoute } from './PrivateRoute';
import { PublicRoute } from './PublicRoute';

export const AppRouter = () => {

    const dispatch = useDispatch();
    // Con esto el appRoutter está pendiente a los cambios de checking y uid (para login y logout)
    const { checking, uid } = useSelector(state => state.auth);

    useEffect(() => {
        // Cada vez que se renderea el componente revisamos si hay token válido
        // y en caso de que lo haya, se renueva
        dispatch(startChecking());
    }, [dispatch]);

    if (checking)
    {
        return <h5>Espere...</h5>;
    }

    const isLogged = !!uid; //!!null === false y !!"asfjlkjfs" === true

    return (
        <Router>
            <div>

                <Switch>
                    <PublicRoute exact path="/login" component={ LoginScreen } isLogged={ isLogged }/>
                    <PrivateRoute exact path="/" component={ CalendarScreen } isLogged={ isLogged }/>
                    <Redirect to="/" />
                </Switch>
                
            </div>
        </Router>
    )
}
