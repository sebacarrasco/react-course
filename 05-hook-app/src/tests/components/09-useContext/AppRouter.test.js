import React from 'react';
import { mount } from 'enzyme';
import { AppRouter } from '../../../components/09-useContext/AppRouter';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('Tests for <AppRouter />', () => {
   
    const user = {
        name: "Sebasti√°n",
        email: "seba@uc.cl"
    }

    const wrapper = mount(
        <UserContext.Provider value={{
            user
        }}>
            <AppRouter />
        </UserContext.Provider>
    );

    test('should match the snapshot', () => {
       
        expect(wrapper).toMatchSnapshot();

    });
    

});
