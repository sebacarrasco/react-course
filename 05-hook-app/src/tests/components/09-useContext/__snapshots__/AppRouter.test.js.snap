// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <AppRouter /> should match the snapshot 1`] = `
<div>
  <nav
    className="navbar navbar-expand-lg navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        aria-current="page"
        className="navbar-brand active"
        href="/"
        onClick={[Function]}
        style={Object {}}
      >
        Usecontext App
      </a>
      <div
        className="collapse navbar-collapse"
        id="navbarNav"
      >
        <div
          className="navbar-nav"
        >
          <a
            aria-current="page"
            className="nav-item nav-link active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Home
          </a>
          <a
            aria-current={null}
            className="nav-item nav-link"
            href="/about"
            onClick={[Function]}
          >
            About
          </a>
          <a
            aria-current={null}
            className="nav-item nav-link"
            href="/login"
            onClick={[Function]}
          >
            Login
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="container"
  >
    <div>
      <h1>
         Home Screen 
      </h1>
      <hr />
      <pre
        className="container"
      >
        {
   "name": "Sebasti√°n",
   "email": "seba@uc.cl"
}
      </pre>
    </div>
  </div>
</div>
`;
